### PR TITLE
Recompile types in a package if its package-info changes

### PR DESCRIPTION
**Pull request by oehme**
Annotation processors can inspect the package of a class,
so we need to make sure that that information is up-to-date
when the package-info changes.

To support this feature we now remember all classes in the
class analysis, not just the ones that had dependents. This
leads to a small increase in memory usage, but means we can
accurately recompile only the classes affected by the given
package change. Alternatively we could omit that data and do
a full recompile instead.

### CODE REVIEW

**Comment by oehme** [comment]
The increase in memory usage is negligible since all the class names are interned. We're only paying the price for the array itself. On our 50K classes project this adds up to 500KB.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathEntrySnapshot.java [hunk 1]
```diff
@@ -91,6 +90,16 @@ private DependentsSet affectedSince(ClasspathEntrySnapshot other) {
                 affected.addAll(dependents.getDependentClasses());
             }
         }
+        for (String added : addedSince(other)) {
+            if (added.endsWith("package-info")) {
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathEntrySnapshot.java]: Can we reference the constant in `ClassSetAnalysisData` or put in in a commonly accessible place?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java [hunk 2]
```diff
@@ -101,12 +124,16 @@ public ClassSetAnalysisData read(Decoder decoder) throws Exception {
...
+            return new ClassSetAnalysisData(classes.build(), dependentsBuilder.build(), classesToConstantsBuilder.build(), classNameToChildren.build(), fullRebuildCause);
         }
 
         @Override
         public void write(Encoder encoder, ClassSetAnalysisData value) throws Exception {
             Map<String, Integer> classNameMap = new HashMap<String, Integer>();
+            encoder.writeSmallInt(value.classes.size());
+            for (String clazz : value.classes) {
+                writeClassName(clazz, classNameMap, encoder);
+            }
```
Comment (marcphilipp) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java]: This is stored in a version-specific cache so we don't have to worry about older versions reading/writing, right?

Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java]: That's right.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathEntrySnapshot.java (hunk 1 above)
Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/ClasspathEntrySnapshot.java]: Will do

**Pull request merged**